main.c

```diff
@@ -274,17 +313,42 @@ bool run_code_py(safe_mode_t safe_mode) {
         filesystem_flush();
         supervisor_allocation* heap = allocate_remaining_memory();
         start_mp(heap);
-        found_main = maybe_run_list(supported_filenames, &result);
-        #if CIRCUITPY_FULL_BUILD
-        if (!found_main){
-            found_main = maybe_run_list(double_extension_filenames, &result);
-            if (found_main) {
-                serial_write_compressed(translate("WARNING: Your code filename has two extensions\n"));
+        uint8_t next_code_options = 0;
+        if (next_code_allocation) {
+            next_code_options = ((next_code_info_t*)next_code_allocation->ptr)->options;
+            const char* next_list[] = {((next_code_info_t*)next_code_allocation->ptr)->filename, ""};
+            found_main = maybe_run_list(next_list, &result);
+            if (!found_main) {
+                serial_write(((next_code_info_t*)next_code_allocation->ptr)->filename);
+                serial_write_compressed(translate(" not found.\n"));
             }
         }
-        #endif
-        cleanup_after_vm(heap);
+        if (!found_main) {
+            found_main = maybe_run_list(supported_filenames, &result);
+            #if CIRCUITPY_FULL_BUILD
+            if (!found_main){
+                found_main = maybe_run_list(double_extension_filenames, &result);
+                if (found_main) {
+                    serial_write_compressed(translate("WARNING: Your code filename has two extensions\n"));
+                }
+            }
+            #endif
+        }
+        bool new_next_code = cleanup_after_vm(heap);
+        // - If a new next code file was set, honor that in any case.
+        // - Otherwise: If a reload was requested by a USB file write, we want to run the same file
+        //   again, preserve any next-code info. Currently that also covers a reload requested by
+        //   supervisor.reload(), or do we want to treat that differently by turning
+        //   reload_requested into an enum?
+        // - Otherwise: Clear any leftover next-code info and restart at code.py/main.py.
```

Comment:
I think it's ok to make it sticky. Save the next code state until it's set back to something else. When on USB, it's ok if it errors. However when trying to do an OTA we may want to go back to code.py or maybe a different default name.

Reply:
Sticky until set doesn’t work well for the “game console” use case where games shouldn’t need to set it to get back to the menu. See https://github.com/adafruit/circuitpython/issues/1084#issuecomment-698851943.